File: zig/src/nucleus/fts.zig
// Nucleus Full-Text Search Model — SQL generation + execution for FTS_* functions
//
// SQL functions: FTS_INDEX, FTS_SEARCH, FTS_FUZZY_SEARCH, FTS_REMOVE,
// FTS_DOC_COUNT, FTS_TERM_COUNT.

const std = @import("std");
const NucleusClient = @import("client.zig").NucleusClient;

pub const FTSModel = struct {
    client: *NucleusClient,

    // ── SQL generators ───────────────────────────────────────────

    pub fn indexSql(doc_id: []const u8, text: []const u8, buf: []u8) ![]const u8 {
        return std.fmt.bufPrint(buf, "SELECT FTS_INDEX('{s}', '{s}')", .{ doc_id, text }) catch return error.BufferTooShort;
    }

    pub fn searchSql(query_str: []const u8, limit: u32, buf: []u8) ![]const u8 {
        return std.fmt.bufPrint(buf, "SELECT FTS_SEARCH('{s}', {d})", .{ query_str, limit }) catch return error.BufferTooShort;
    }

    pub fn fuzzySearchSql(query_str: []const u8, max_distance: u32, limit: u32, buf: []u8) ![]const u8 {
        return std.fmt.bufPrint(buf, "SELECT FTS_FUZZY_SEARCH('{s}', {d}, {d})", .{ query_str, max_distance, limit }) catch return error.BufferTooShort;
    }

    pub fn removeSql(doc_id: []const u8, buf: []u8) ![]const u8 {
        return std.fmt.bufPrint(buf, "SELECT FTS_REMOVE('{s}')", .{doc_id}) catch return error.BufferTooShort;
    }

    pub fn docCountSql(buf: []u8) ![]const u8 {
        return std.fmt.bufPrint(buf, "SELECT FTS_DOC_COUNT()", .{}) catch return error.BufferTooShort;
    }

    pub fn termCountSql(buf: []u8) ![]const u8 {
        return std.fmt.bufPrint(buf, "SELECT FTS_TERM_COUNT()", .{}) catch return error.BufferTooShort;
    }

    // ── Execution methods ────────────────────────────────────────

    pub fn ftsIndex(self: FTSModel, doc_id: []const u8, text: []const u8) !?[]const u8 {
        var buf: [4096]u8 = undefined;
        const sql = try indexSql(doc_id, text, &buf);
        return try self.client.execute(sql);
    }

    pub fn search(self: FTSModel, query_str: []const u8, limit: u32) !?[]const u8 {
        var buf: [512]u8 = undefined;
        const sql = try searchSql(query_str, limit, &buf);
        return try self.client.execute(sql);
    }

    pub fn fuzzySearch(self: FTSModel, query_str: []const u8, max_distance: u32, limit: u32) !?[]const u8 {
        var buf: [512]u8 = undefined;
        const sql = try fuzzySearchSql(query_str, max_distance, limit, &buf);
        return try self.client.execute(sql);
    }

    pub fn remove(self: FTSModel, doc_id: []const u8) !?[]const u8 {
        var buf: [256]u8 = undefined;
        const sql = try removeSql(doc_id, &buf);
        return try self.client.execute(sql);
    }

    pub fn docCount(self: FTSModel) !?[]const u8 {
        var buf: [256]u8 = undefined;
        const sql = try docCountSql(&buf);
        return try self.client.execute(sql);
    }

    pub fn termCount(self: FTSModel) !?[]const u8 {
        var buf: [256]u8 = undefined;
        const sql = try termCountSql(&buf);
        return try self.client.execute(sql);
    }
};

// ── Tests ─────────────────────────────────────────────────────

test "FTS_INDEX sql" {
    var buf: [512]u8 = undefined;
    const sql = try FTSModel.indexSql("doc-42", "The quick brown fox jumps over the lazy dog", &buf);
    try std.testing.expectEqualStrings("SELECT FTS_INDEX('doc-42', 'The quick brown fox jumps over the lazy dog')", sql);
}

test "FTS_SEARCH sql" {
    var buf: [256]u8 = undefined;
    const sql = try FTSModel.searchSql("brown fox", 20, &buf);
    try std.testing.expectEqualStrings("SELECT FTS_SEARCH('brown fox', 20)", sql);
}

test "FTS_FUZZY_SEARCH sql" {
    var buf: [256]u8 = undefined;
    const sql = try FTSModel.fuzzySearchSql("quik", 2, 10, &buf);
    try std.testing.expectEqualStrings("SELECT FTS_FUZZY_SEARCH('quik', 2, 10)", sql);
}

test "FTS_REMOVE sql" {
    var buf: [256]u8 = undefined;
    const sql = try FTSModel.removeSql("doc-42", &buf);
    try std.testing.expectEqualStrings("SELECT FTS_REMOVE('doc-42')", sql);
}

test "FTS_DOC_COUNT sql" {
    var buf: [256]u8 = undefined;
    const sql = try FTSModel.docCountSql(&buf);
    try std.testing.expectEqualStrings("SELECT FTS_DOC_COUNT()", sql);
}

test "FTS_TERM_COUNT sql" {
    var buf: [256]u8 = undefined;
    const sql = try FTSModel.termCountSql(&buf);
    try std.testing.expectEqualStrings("SELECT FTS_TERM_COUNT()", sql);
}

File: elixir/lib/nucleus/models/fts.ex
defmodule Nucleus.Models.FTS do
  @moduledoc """
  Full-text search model — FTS_INDEX, FTS_SEARCH, FTS_FUZZY_SEARCH.

  ## Example

      alias Nucleus.Models.FTS

      FTS.index(client, 1, "The quick brown fox")
      FTS.index(client, 2, "The lazy dog")

      {:ok, results} = FTS.search(client, "quick fox", 10)
      {:ok, fuzzy} = FTS.fuzzy_search(client, "quik", 2, 10)
  """

  @type client :: Nucleus.Client.t()

  @doc "Indexes text content for a document ID."
  @spec index(client(), integer(), String.t()) :: {:ok, boolean()} | {:error, term()}
  def index(client, doc_id, text) do
    with :ok <- Nucleus.Client.require_nucleus(client, "FTS.index") do
      case Nucleus.Client.query(client, "SELECT FTS_INDEX($1, $2)", [doc_id, text]) do
        {:ok, %{rows: [[val]]}} -> {:ok, val}
        {:error, _} = error -> error
      end
    end
  end

  @doc """
  Searches indexed text. Returns a list of `%{doc_id: id, score: score}` results.
  """
  @spec search(client(), String.t(), integer()) ::
          {:ok, [%{doc_id: integer(), score: float()}]} | {:error, term()}
  def search(client, query, limit \\ 10) do
    with :ok <- Nucleus.Client.require_nucleus(client, "FTS.search") do
      case Nucleus.Client.query(client, "SELECT FTS_SEARCH($1, $2)", [query, limit]) do
        {:ok, %{rows: [[json]]}} when is_binary(json) ->
          {:ok, Jason.decode!(json) |> normalize_results()}

        {:ok, _} ->
          {:ok, []}

        {:error, _} = error ->
          error
      end
    end
  end

  @doc """
  Fuzzy search with a maximum edit distance.
  """
  @spec fuzzy_search(client(), String.t(), integer(), integer()) ::
          {:ok, [%{doc_id: integer(), score: float()}]} | {:error, term()}
  def fuzzy_search(client, query, max_distance, limit \\ 10) do
    with :ok <- Nucleus.Client.require_nucleus(client, "FTS.fuzzy_search") do
      case Nucleus.Client.query(client, "SELECT FTS_FUZZY_SEARCH($1, $2, $3)", [
             query,
             max_distance,
             limit
           ]) do
        {:ok, %{rows: [[json]]}} when is_binary(json) ->
          {:ok, Jason.decode!(json) |> normalize_results()}

        {:ok, _} ->
          {:ok, []}

        {:error, _} = error ->
          error
      end
    end
  end

  @doc "Removes a document from the search index."
  @spec remove(client(), integer()) :: {:ok, boolean()} | {:error, term()}
  def remove(client, doc_id) do
    with :ok <- Nucleus.Client.require_nucleus(client, "FTS.remove") do
      case Nucleus.Client.query(client, "SELECT FTS_REMOVE($1)", [doc_id]) do
        {:ok, %{rows: [[val]]}} -> {:ok, val}
        {:error, _} = error -> error
      end
    end
  end

  @doc "Returns the total number of indexed documents."
  @spec doc_count(client()) :: {:ok, integer()} | {:error, term()}
  def doc_count(client) do
    with :ok <- Nucleus.Client.require_nucleus(client, "FTS.doc_count") do
      case Nucleus.Client.query(client, "SELECT FTS_DOC_COUNT()", []) do
        {:ok, %{rows: [[val]]}} -> {:ok, val}
        {:error, _} = error -> error
      end
    end
  end

  @doc "Returns the total number of indexed terms."
  @spec term_count(client()) :: {:ok, integer()} | {:error, term()}
  def term_count(client) do
    with :ok <- Nucleus.Client.require_nucleus(client, "FTS.term_count") do
      case Nucleus.Client.query(client, "SELECT FTS_TERM_COUNT()", []) do
        {:ok, %{rows: [[val]]}} -> {:ok, val}
        {:error, _} = error -> error
      end
    end
  end

  defp normalize_results(results) when is_list(results) do
    Enum.map(results, fn item ->
      %{
        doc_id: item["doc_id"] || item["id"],
        score: item["score"] || 0.0
      }
    end)
  end

  defp normalize_results(_), do: []
end
